### PR TITLE
Add support for proxy in local provider

### PR DESCRIPTION
**Pull request by hemanthnakkina**
* Add default proxies in local provider based on entries in /etc/environment.
* Add 127.0.0.1, localhost, svc, k8s cluster service cidr, k8s cluster pod cidr to no_proxy
* Add step to prompt user for proxies
* Update juju bootstrap command based on proxies.
* Pass proxies as env to process running juju bootstrap command
* Update sunbeam-machine charm proxy configs as part of application deployment
* Pass proxies to terraform helper as part of environment so that proxies will be used in running terraform commands
* Add checks in prepare-node script to find out if proxies need to be set by user.

### CODE REVIEW

File: sunbeam-python/sunbeam/jobs/steps.py [hunk 1]
```diff
@@ -322,3 +330,98 @@ def run(self, status: Optional[Status] = None) -> Result:
             return Result(ResultType.FAILED, str(e))
 
         return Result(ResultType.COMPLETED)
+
+
+def proxy_questions():
+    return {
+        "proxy_required": PromptQuestion(
```
Comment (gboutry) [inline sunbeam-python/sunbeam/jobs/steps.py:337]: You can use a `ConfirmQuestion` not to handle the y/n specifically here

File: sunbeam-python/sunbeam/jobs/common.py [hunk 2]
```diff
@@ -389,3 +395,36 @@ def str_presenter(dumper: yaml.Dumper, data: str) -> yaml.ScalarNode:
...
+def get_proxy_settings(deployment: Deployment) -> dict:
+    proxy = {}
+    client = deployment.get_client()
+    try:
+        proxy_from_db = read_config(client, PROXY_CONFIG_KEY) or {}
+        proxy = {
+            p.upper(): v
+            for p in ("http_proxy", "https_proxy", "no_proxy")
+            if (v := proxy_from_db.get("proxy", {}).get(p))
+        }
```
Comment (gboutry) [inline sunbeam-python/sunbeam/jobs/common.py:420]: That's a bit hard to read:
```suggestion
        proxy_from_db = read_config(client, PROXY_CONFIG_KEY).get("proxy")
        proxy = {
            p.upper(): v
            for p in ("http_proxy", "https_proxy", "no_proxy")
            if (v := proxy_from_db.get(p))
        }
```

File: sunbeam-python/sunbeam/jobs/common.py [hunk 3]
```diff
@@ -389,3 +395,36 @@ def str_presenter(dumper: yaml.Dumper, data: str) -> yaml.ScalarNode:
...
+    except (ClusterServiceUnavailableException, ConfigItemNotFoundException) as e:
+        LOG.debug(f"Using default Proxy settings from provider due to {str(e)}")
+        proxy = deployment.get_default_proxy_settings()
+
+    if "NO_PROXY" in proxy:
+        no_proxy_list = set(proxy.get("NO_PROXY").split(","))
+        missing_no_proxies = _get_default_no_proxy_settings() - no_proxy_list
+        proxy["NO_PROXY"] = ",".join(no_proxy_list.union(missing_no_proxies))
```
Comment (gboutry) [inline sunbeam-python/sunbeam/jobs/common.py:428]: Computing missing + union is unnecessary, or am I missing the point ?
Set is already taking care of deduplicating
```suggestion
        no_proxies =  _get_default_no_proxy_settings().union(proxy.get("NO_PROXY", "").split(","))
        proxy["NO_PROXY"] = ",".join(no_proxies)
```

**Review by gboutry — CHANGES_REQUESTED** [review]

File: sunbeam-python/sunbeam/jobs/steps.py (hunk 1 above)
Comment (hemanthnakkina) [reply sunbeam-python/sunbeam/jobs/steps.py]: Done

File: sunbeam-python/sunbeam/jobs/common.py (hunk 2 above)
Comment (hemanthnakkina) [reply sunbeam-python/sunbeam/jobs/common.py]: Done

File: sunbeam-python/sunbeam/jobs/common.py (hunk 3 above)
Comment (hemanthnakkina) [reply sunbeam-python/sunbeam/jobs/common.py]: Done

**Review by gboutry — APPROVED** [review]

**Pull request merged**